freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/EnvironmentUserSyncStateCalculator.java

```diff
@@ -0,0 +1,115 @@
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.UserSyncState;
+import com.sequenceiq.freeipa.entity.Operation;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.entity.UserSyncStatus;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UmsEventGenerationIds;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Service
+public class EnvironmentUserSyncStateCalculator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EnvironmentUserSyncStateCalculator.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private UserSyncStatusService userSyncStatusService;
+
+    @Inject
+    private UmsEventGenerationIdsProvider umsEventGenerationIdsProvider;
+
+    @Inject
+    private EventGenerationIdsChecker eventGenerationIdsChecker;
+
+    public EnvironmentUserSyncState calculateEnvironmentUserSyncState(String accountId, Crn environmentCrn) {
+        checkArgument(accountId.equals(environmentCrn.getAccountId()), "environmentCrn does not match account id");
+        String envCrnString = environmentCrn.toString();
+
+        Stack stack = stackService.getByEnvironmentCrnAndAccountId(envCrnString, accountId);
+        MDCBuilder.buildMdcContext(stack);
+
+        UserSyncStatus userSyncStatus = userSyncStatusService.findByStack(stack);
+
+        return internalCalculateEnvironmentUserSyncState(accountId, envCrnString, userSyncStatus);
+    }
+
+    @VisibleForTesting
+    EnvironmentUserSyncState internalCalculateEnvironmentUserSyncState(String accountId, String envCrnString, UserSyncStatus userSyncStatus) {
+        EnvironmentUserSyncState environmentUserSyncState = new EnvironmentUserSyncState();
+        if (null == userSyncStatus || null == userSyncStatus.getLastStartedFullSync()) {
+            environmentUserSyncState.setState(UserSyncState.STALE);
+        } else {
+            environmentUserSyncState.setLastUserSyncOperationId(userSyncStatus.getLastStartedFullSync().getOperationId());
+            environmentUserSyncState.setState(calculateUserSyncState(accountId, envCrnString, userSyncStatus));
+        }
+
+        return environmentUserSyncState;
+    }
+
+    private UserSyncState calculateUserSyncState(String accountId, String envCrnString, UserSyncStatus userSyncStatus) {
+        Operation lastSync = userSyncStatus.getLastStartedFullSync();
+        UserSyncState state;
+        switch (lastSync.getStatus()) {
+            case RUNNING:
+                state = UserSyncState.SYNC_IN_PROGRESS;
+                break;
+            case COMPLETED:
+                state = calculateStateForCompletedOperation(accountId, envCrnString, userSyncStatus);
+                break;
+            case REQUESTED:
+            case REJECTED:
+                // REQUESTED or REJECTED operations will never be saved as part of the UserSyncStatus
+                state = calculateStateForUnexpectedOperationStatus(envCrnString, userSyncStatus);
```

Comment:
this name is a bit misleading as I would expect it would set the state, but instead it will throw an exception. also this method should be void as it won't return anything ever.

Reply:
I renamed the method but I also changed it to return the. the case statement needs to throw the exception instead of the method. Otherwise, the compiler can't tell that the method will always throw and will complain about some codepaths not setting the `state` variable.

Reply:
thanks for the change!